samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,379 @@
+
+    public TriggerContextImpl(TriggerKey<K> triggerKey) {
+      this.triggerKey = triggerKey;
+    }
+
+    @Override
+    public Cancellable scheduleCallback(Runnable runnable, long callbackTimeMs) {
+      TriggerCallbackState timerState = new TriggerCallbackState(triggerKey, runnable, callbackTimeMs);
+      pendingCallbacks.add(timerState);
+      return timerState;
+    }
+  }
+
+  /**
+   * State corresponding to pending timer callbacks scheduled by various {@link TriggerImpl}s.
+   */
+  private class TriggerCallbackState implements Comparable<TriggerCallbackState>, Cancellable {
+
+    private final TriggerKey<K> triggerKey;
+    private final Runnable callback;
+
+    //the time in milliseconds at which the callback should trigger
+    private final long scheduleTimeMs;
+
+    private TriggerCallbackState(TriggerKey<K> triggerKey, Runnable callback, long scheduleTimeMs) {
+      this.triggerKey = triggerKey;
+      this.callback = callback;
+      this.scheduleTimeMs = scheduleTimeMs;
+    }
+
+    private Runnable getCallback() {
+      return callback;
+    }
+
+    private long getScheduleTimeMs() {
+      return scheduleTimeMs;
+    }
+
+    @Override
+    public int compareTo(TriggerCallbackState other) {
+      return Long.compare(this.scheduleTimeMs, other.scheduleTimeMs);
+    }
+
+    @Override
+    public boolean cancel() {
+      return pendingCallbacks.remove(this);
+    }
+  }
+
+  /**
+   * State corresponding to a created {@link TriggerImpl} instance.
+   */
+  private class TriggerImplState {
+    private final TriggerKey<K> triggerKey;
+    // The context, and the {@link TriggerImpl} instance corresponding to this triggerKey
+    private final TriggerImpl<M> impl;
+    private final TriggerContext context;
+    // Guard to ensure that we don't invoke onMessage on already cancelled triggers
+    private boolean isCancelled = false;
+
+    public TriggerImplState(TriggerKey<K> triggerKey, TriggerImpl<M> impl, TriggerContext context) {
+      this.triggerKey = triggerKey;
+      this.impl = impl;
+      this.context = context;
+    }
+
+    public void onMessage(M message, MessageCollector collector, TaskCoordinator coordinator) {
+      if (!isCancelled) {
+        impl.onMessage(message, context);
+
+        if (impl.shouldFire()) {
+          //repeating trigger can trigger multiple times, So, clear the state to allow future triggerings.
+          impl.clear();
+          onTriggerFired(triggerKey, collector, coordinator);
+        }
+      }
+    }
+
+    public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+      if (impl.shouldFire() && !isCancelled) {
+        //repeating trigger can trigger multiple times, So, clear the trigger to allow future triggerings.
+        impl.clear();
+        onTriggerFired(triggerKey, collector, coordinator);
+      }
+    }
+
+    public void cancel() {
+      impl.cancel();
+      isCancelled = true;
+    }
+
+    public boolean isRepeating() {
```

Comment:
I don't quite like the name wrapper.... It does not tell me what this class is supposed to do. In fact, TriggerImplState sounds better to me because this class does keep all relevant runtime objects and states related to a trigger. If the reason for renaming is simply TriggerImplState.isRepeating() sounds weird, we can change the method to TriggerImplState.canRepeat().